src/gt4py/next/iterator/transforms/cse.py

```diff
@@ -92,24 +101,29 @@ def remove_subexprs(self, nodes: typing.Iterable[ir.Node]) -> None:
             for node in nodes:
                 subexpr_data = self.subexprs.pop(node, None)
                 if subexpr_data:
-                    node_ids, _ = zip(*subexpr_data)
+                    node_ids, _ = zip(*subexpr_data.subexprs)
                     node_ids_to_remove |= set(node_ids)
             for subexpr_data in self.subexprs.values():
-                for _, collected_child_node_ids in subexpr_data:
+                for _, collected_child_node_ids in subexpr_data.subexprs:
                     collected_child_node_ids -= node_ids_to_remove
 
     @classmethod
     def apply(cls, node: ir.Node) -> dict[ir.Node, list[tuple[int, set[int]]]]:
         state = cls.State()
         obj = cls()
-        obj.visit(node, state=state)
-        # return subexpression in pre-order of the tree, i.e. the nodes closer to the root come
-        # first, and skip the root node itself
-        return {k: v for k, v in reversed(state.subexprs.items()) if k is not node}
+        obj.visit(node, state=state, depth=-1)
+        # Return subexpression such that the nodes closer to the root come first and skip the root
+        # node itself.
+        subexprs_sorted: list[tuple[ir.Node, "CollectSubexpressions.SubexpressionData"]] = sorted(
+            state.subexprs.items(), key=lambda el: el[1].max_depth
+        )
+        return {k: v.subexprs for k, v in subexprs_sorted if k is not node}
 
     def visit(self, node: ir.Node, **kwargs) -> None:  # type: ignore[override]  # supertype accepts any node, but we want to be more specific here.
+        depth = kwargs.pop("depth")
+
         if not isinstance(node, SymbolTableTrait) and not _is_collectable_expr(node):
-            return super().visit(node, **kwargs)
+            return super().visit(node, depth=depth + 1, **kwargs)
```

Comment:
does this make sense? we are visiting the same node, but call the visit of the super class?

Reply:
I agree this is not nice. Except for the `visit_SymRef` case `super().visit` will dispatch to `generic_visit` and visit the children where `depth + 1` is correct. I'm open for other solutions.

Reply:
Would it make sense to call generic_visit directly?

Reply:
overload generic_visit and increase depth there.